helix-term/src/ui/statusline.rs

```diff
@@ -164,162 +155,130 @@ where
     }
 }
 
-fn render_mode<F>(context: &mut RenderContext, write: F)
-where
-    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
-{
-    let visible = context.focused;
+fn render_mode<'a>(context: &RenderContext) -> Spans<'a> {
     let config = context.editor.config();
     let modenames = &config.statusline.mode;
-    write(
-        context,
-        format!(
-            " {} ",
-            if visible {
+
+    let mode_str = match context.editor.mode() {
+        Mode::Normal => &modenames.normal,
+        Mode::Select => &modenames.select,
+        Mode::Insert => &modenames.insert,
+    };
+
+    if context.focused {
+        if config.color_modes {
+            Spans::single(
+                format!(" {} ", mode_str),
                 match context.editor.mode() {
-                    Mode::Insert => &modenames.insert,
-                    Mode::Select => &modenames.select,
-                    Mode::Normal => &modenames.normal,
-                }
-            } else {
-                // If not focused, explicitly leave an empty space instead of returning None.
-                "   "
-            }
-        ),
-        if visible && config.color_modes {
-            match context.editor.mode() {
-                Mode::Insert => Some(context.editor.theme.get("ui.statusline.insert")),
-                Mode::Select => Some(context.editor.theme.get("ui.statusline.select")),
-                Mode::Normal => Some(context.editor.theme.get("ui.statusline.normal")),
-            }
+                    Mode::Normal => context.editor.theme.get("ui.statusline.normal"),
+                    Mode::Select => context.editor.theme.get("ui.statusline.select"),
+                    Mode::Insert => context.editor.theme.get("ui.statusline.insert"),
+                },
+            )
         } else {
-            None
-        },
-    );
+            Spans::single(String::from(mode_str), context.base_style)
+        }
+    } else {
+        Spans::new()
+    }
 }
 
 // TODO think about handling multiple language servers
-fn render_lsp_spinner<F>(context: &mut RenderContext, write: F)
-where
-    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
-{
-    let language_server = context.doc.language_servers().next();
-    write(
-        context,
-        language_server
-            .and_then(|srv| {
-                context
-                    .spinners
-                    .get(srv.id())
-                    .and_then(|spinner| spinner.frame())
-            })
-            // Even if there's no spinner; reserve its space to avoid elements frequently shifting.
-            .unwrap_or(" ")
```

Comment:
But if the behavior is kept then there will be a large empty space between the elements surrounding the spinner when the spinner is not in use.

I guess either situation could be solved by simply placing the spinner as the last item of the left side or the first item of the right side, blending it in with the surrounding spacing. Or we could change the spinner idle frame to always display a faded out placeholder when not in use, e.g. `⣿`.

Reply:
Items already jump around for other reasons, right? Almost very single element that can go in the status line can cause shifts (diagnostics, selections, position, etc.).

I simply put the elements that cause the most shift or most frequent shift (diagnostics) at the end of the list.

Reply:
No feedback arrived for a while on this, and since I agree with @twharmon I will mark this as resolved for now. 

As I said, if the spinner absolutely _must_ reserve space we should use a visible idle character, but it would be more elegant to simply place the spinner on the inner edge of the side section where it won't push other elements around.